File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/android/core/helpers/BildirimRolPolitikasi.kt
package com.satinalmapro.android.core.helpers

import com.satinalmapro.android.core.model.BildirimRecord
import com.satinalmapro.android.core.model.BildirimTipleri
import com.satinalmapro.android.core.model.UserProfile
import com.satinalmapro.android.core.roles.KullaniciRolleri

/** Yönetim yalnızca: saha talebi + teklif onayı bildirimlerini görür. */
object BildirimRolPolitikasi {
    fun islemYapanKendisiMi(record: BildirimRecord, user: UserProfile?): Boolean =
        user != null &&
            record.olusturanUid.isNotBlank() &&
            record.olusturanUid.equals(user.uid, ignoreCase = true)

    /** WhatsApp/Instagram kuralı: işlemi yapan kişiye kayıt/push oluşturulmaz. */
    fun kayitGonderilmeli(hedefRol: String?, hedefUid: String?, islemYapanUid: String?): Boolean {
        if (islemYapanUid.isNullOrBlank())
            return !hedefRol.isNullOrBlank() || !hedefUid.isNullOrBlank()
        if (!hedefUid.isNullOrBlank() && hedefUid.equals(islemYapanUid, ignoreCase = true))
            return false
        return !hedefRol.isNullOrBlank() || !hedefUid.isNullOrBlank()
    }

    fun kayitGonderilmeli(record: BildirimRecord): Boolean =
        kayitGonderilmeli(record.hedefRol, record.hedefUid, record.olusturanUid)

    fun rolTipGorebilirMi(rol: String?, tip: String): Boolean {
        if (KullaniciRolleri.isAdmin(rol)) return true
        val r = KullaniciRolleri.normalize(rol)
        val t = normalizeTip(tip)
        return when (t) {
            BildirimTipleri.YONETIME_GONDERILDI ->
                r == KullaniciRolleri.YONETIM
            BildirimTipleri.TEKLIF_ONAYDA -> r == KullaniciRolleri.YONETIM
            BildirimTipleri.TEKLIF_ISTENDI, BildirimTipleri.TEKLIF_DUZELTME_ISTENDI ->
                r == KullaniciRolleri.SATINALMA
            BildirimTipleri.ONAYLANDI -> r == KullaniciRolleri.SATINALMA
            BildirimTipleri.REDDEDILDI -> r == KullaniciRolleri.SATINALMA
            BildirimTipleri.SIPARIS_OLUSTURULDU ->
                r == KullaniciRolleri.SATINALMA || r == KullaniciRolleri.DEPO || r == KullaniciRolleri.ATOLYE
            BildirimTipleri.MAL_KABUL_EDILDI -> r == KullaniciRolleri.SATINALMA
            else -> false
        }
    }

    fun kullaniciyaMi(record: BildirimRecord, user: UserProfile?): Boolean {
        if (user == null) return false
        if (islemYapanKendisiMi(record, user)) return false

        if (!record.hedefUid.isNullOrBlank()) {
            return record.hedefUid.equals(user.uid, ignoreCase = true)
        }

        if (!record.inboxDocId.isNullOrBlank()) {
            return rolTipGorebilirMi(user.role, record.tip)
        }

        if (KullaniciRolleri.isAdmin(user.role)) return true

        if (!record.hedefRol.isNullOrBlank()) {
            val rol = KullaniciRolleri.normalize(user.role)
            return KullaniciRolleri.normalize(record.hedefRol) == rol &&
                rolTipGorebilirMi(rol, record.tip)
        }

        return false
    }

    fun yonetimeGonderildiHedefleri(): List<Pair<String?, String?>> =
        listOf(KullaniciRolleri.YONETIM to null)

    fun reddedildiHedefleri(talepOlusturanUid: String?, islemYapanUid: String?): List<Pair<String?, String?>> {
        val hedefler = mutableListOf<Pair<String?, String?>>()
        hedefler += KullaniciRolleri.SATINALMA to null
        if (!talepOlusturanUid.isNullOrBlank() &&
            !talepOlusturanUid.equals(islemYapanUid, ignoreCase = true)
        ) {
            hedefler += null to talepOlusturanUid
        }
        return hedefler
    }

    fun siparisOlusturulduHedefleri(talepOlusturanUid: String?, islemYapanUid: String?): List<Pair<String?, String?>> {
        val hedefler = mutableListOf<Pair<String?, String?>>()
        hedefler += KullaniciRolleri.SATINALMA to null
        hedefler += KullaniciRolleri.DEPO to null
        hedefler += KullaniciRolleri.ATOLYE to null
        if (!talepOlusturanUid.isNullOrBlank() &&
            !talepOlusturanUid.equals(islemYapanUid, ignoreCase = true)
        ) {
            hedefler += null to talepOlusturanUid
        }
        return hedefler
    }

    fun malKabulEdildiHedefleri(talepOlusturanUid: String?, islemYapanUid: String?): List<Pair<String?, String?>> {
        val hedefler = mutableListOf<Pair<String?, String?>>()
        hedefler += KullaniciRolleri.SATINALMA to null
        if (!talepOlusturanUid.isNullOrBlank() &&
            !talepOlusturanUid.equals(islemYapanUid, ignoreCase = true)
        ) {
            hedefler += null to talepOlusturanUid
        }
        return hedefler
    }

    private fun normalizeTip(tip: String): String = when (tip.lowercase()) {
        "yonetimegonderildi" -> BildirimTipleri.YONETIME_GONDERILDI
        "teklifistendi" -> BildirimTipleri.TEKLIF_ISTENDI
        "teklifonayda" -> BildirimTipleri.TEKLIF_ONAYDA
        "teklifduzeltmeistendi" -> BildirimTipleri.TEKLIF_DUZELTME_ISTENDI
        "onaylandi" -> BildirimTipleri.ONAYLANDI
        "reddedildi" -> BildirimTipleri.REDDEDILDI
        "siparisolusturuldu" -> BildirimTipleri.SIPARIS_OLUSTURULDU
        "malkabuledildi" -> BildirimTipleri.MAL_KABUL_EDILDI
        else -> tip.lowercase()
    }
}

File: SatinalmaPro.Shared/Helpers/BildirimRolPolitikasi.cs
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Helpers;

/// <summary>
/// Rol × bildirim tipi matrisi — masaüstü, mobil ve Android aynı kuralları kullanır.
/// Talep/teklif onay bildirimleri: Yönetim + Satınalma (aynı karar yetkisi).
/// </summary>
public static class BildirimRolPolitikasi
{
    public static bool IslemYapanKendisiMi(BildirimKaydi bildirim, KullaniciProfili? kullanici)
    {
        if (kullanici is null || string.IsNullOrWhiteSpace(kullanici.Uid))
            return false;
        if (!string.IsNullOrWhiteSpace(bildirim.OlusturanUid)
            && string.Equals(bildirim.OlusturanUid, kullanici.Uid, StringComparison.OrdinalIgnoreCase))
            return true;
        return false;
    }

    /// <summary>WhatsApp/Instagram kuralı: işlemi yapan kişiye kayıt/push oluşturulmaz.</summary>
    public static bool KayitGonderilmeli(string? hedefRol, string? hedefUid, string? islemYapanUid)
    {
        if (string.IsNullOrWhiteSpace(islemYapanUid))
            return !string.IsNullOrWhiteSpace(hedefRol) || !string.IsNullOrWhiteSpace(hedefUid);

        if (!string.IsNullOrWhiteSpace(hedefUid)
            && string.Equals(hedefUid, islemYapanUid, StringComparison.OrdinalIgnoreCase))
            return false;

        return !string.IsNullOrWhiteSpace(hedefRol) || !string.IsNullOrWhiteSpace(hedefUid);
    }

    public static bool KayitGonderilmeli(BildirimKaydi bildirim) =>
        KayitGonderilmeli(bildirim.HedefRol, bildirim.HedefUid, bildirim.OlusturanUid);

    /// <summary>Admin tüm akış bildirimlerini görür.</summary>
    public static bool RolTipGorebilirMi(string? rol, string tip)
    {
        if (KullaniciRolleri.AdminMi(rol))
            return true;

        var r = KullaniciRolleri.Normalize(rol);
        tip = NormalizeTip(tip);

        return tip switch
        {
            BildirimTipleri.YonetimeGonderildi =>
                r is KullaniciRolleri.Yonetim or KullaniciRolleri.Satinalma,
            BildirimTipleri.TeklifOnayda =>
                r is KullaniciRolleri.Yonetim or KullaniciRolleri.Satinalma,
            BildirimTipleri.TeklifIstendi or BildirimTipleri.TeklifDuzeltmeIstendi =>
                r == KullaniciRolleri.Satinalma,
            BildirimTipleri.Onaylandi =>
                r == KullaniciRolleri.Satinalma,
            BildirimTipleri.Reddedildi =>
                r == KullaniciRolleri.Satinalma,
            BildirimTipleri.SiparisOlusturuldu =>
                r is KullaniciRolleri.Satinalma or KullaniciRolleri.Depo,
            BildirimTipleri.MalKabulEdildi =>
                r == KullaniciRolleri.Satinalma,
            _ => false
        };
    }

    public static bool KullaniciyaMi(BildirimKaydi bildirim, KullaniciProfili? kullanici)
    {
        if (kullanici is null)
            return false;

        // İşlemi yapan kişi kendi aksiyonunun bildirimini asla almaz.
        if (IslemYapanKendisiMi(bildirim, kullanici))
            return false;

        if (!string.IsNullOrWhiteSpace(bildirim.HedefUid))
            return bildirim.HedefUid == kullanici.Uid;

        if (!string.IsNullOrWhiteSpace(bildirim.InboxDocId))
            return RolTipGorebilirMi(kullanici.Rol, bildirim.Tip);

        if (KullaniciRolleri.AdminMi(kullanici.Rol))
            return true;

        if (!string.IsNullOrWhiteSpace(bildirim.HedefRol))
        {
            var rol = KullaniciRolleri.Normalize(kullanici.Rol);
            return KullaniciRolleri.Normalize(bildirim.HedefRol) == rol
                && RolTipGorebilirMi(rol, bildirim.Tip);
        }

        return false;
    }

    public static IEnumerable<(string? HedefRol, string? HedefUid)> YonetimeGonderildiHedefleri() =>
    [
        (KullaniciRolleri.Yonetim, null),
        (KullaniciRolleri.Satinalma, null)
    ];

    public static IEnumerable<(string? HedefRol, string? HedefUid)> TeklifOnaydaHedefleri() =>
    [
        (KullaniciRolleri.Yonetim, null),
        (KullaniciRolleri.Satinalma, null)
    ];

    public static IEnumerable<(string? HedefRol, string? HedefUid)> ReddedildiHedefleri(
        string? talepOlusturanUid,
        string? islemYapanUid)
    {
        yield return (KullaniciRolleri.Satinalma, null);
        if (!string.IsNullOrWhiteSpace(talepOlusturanUid)
            && !string.Equals(talepOlusturanUid, islemYapanUid, StringComparison.OrdinalIgnoreCase))
            yield return (null, talepOlusturanUid);
    }

    public static IEnumerable<(string? HedefRol, string? HedefUid)> SiparisOlusturulduHedefleri(
        string? talepOlusturanUid,
        string? islemYapanUid)
    {
        yield return (KullaniciRolleri.Satinalma, null);
        yield return (KullaniciRolleri.Depo, null);
        if (!string.IsNullOrWhiteSpace(talepOlusturanUid)
            && !string.Equals(talepOlusturanUid, islemYapanUid, StringComparison.OrdinalIgnoreCase))
            yield return (null, talepOlusturanUid);
    }

    public static IEnumerable<(string? HedefRol, string? HedefUid)> MalKabulEdildiHedefleri(
        string? talepOlusturanUid,
        string? islemYapanUid)
    {
        yield return (KullaniciRolleri.Satinalma, null);
        if (!string.IsNullOrWhiteSpace(talepOlusturanUid)
            && !string.Equals(talepOlusturanUid, islemYapanUid, StringComparison.OrdinalIgnoreCase))
            yield return (null, talepOlusturanUid);
    }

    public static string NormalizeTip(string tip)
    {
        var t = tip.Trim().ToLowerInvariant();
        // Event kodları (talep.olusturuldu …)
        t = t switch
        {
            "talep.yonetime_gonderildi" or "talep.olusturuldu"
                or "talep.sla_yaklasiyor" or "talep.sla_asildi" => BildirimTipleri.YonetimeGonderildi,
            "teklif.istendi" => BildirimTipleri.TeklifIstendi,
            "teklif.yonetime_gonderildi" => BildirimTipleri.TeklifOnayda,
            "teklif.duzeltme_istendi" => BildirimTipleri.TeklifDuzeltmeIstendi,
            "talep.onaylandi" => BildirimTipleri.Onaylandi,
            "talep.reddedildi" => BildirimTipleri.Reddedildi,
            "siparis.olusturuldu" => BildirimTipleri.SiparisOlusturuldu,
            "depo.mal_kabul_yapildi" => BildirimTipleri.MalKabulEdildi,
            _ => t
        };

        return t switch
        {
            "yonetimegonderildi" or "yonetime_gonderildi" => BildirimTipleri.YonetimeGonderildi,
            "teklifistendi" or "teklif_istendi" => BildirimTipleri.TeklifIstendi,
            "teklifonayda" or "teklif_onayda" => BildirimTipleri.TeklifOnayda,
            "teklifduzeltmeistendi" or "teklif_duzeltme_istendi" => BildirimTipleri.TeklifDuzeltmeIstendi,
            "onaylandi" => BildirimTipleri.Onaylandi,
            "reddedildi" => BildirimTipleri.Reddedildi,
            "siparisolusturuldu" or "siparis_olusturuldu" => BildirimTipleri.SiparisOlusturuldu,
            "malkabuledildi" or "mal_kabul_edildi" => BildirimTipleri.MalKabulEdildi,
            _ => t
        };
    }
}
